### PR TITLE
[maskedtensor] Sparsity tutorial [2/4]

### PR DESCRIPTION
**Pull request by george-qi**
Continuation from https://github.com/pytorch/tutorials/pull/2050

The new format will be (below are the previous PRs):

PR #1: Overview Tutorial
- [Old overview](https://github.com/pytorch/tutorials/pull/2042)
- Why is MaskedTensor useful
    - [Distinguish 0 and NaN gradient](https://github.com/pytorch/tutorials/pull/2044)
    - [Safe softmax tutorial](https://github.com/pytorch/tutorials/pull/2045)
    - [Missing nan ops](https://github.com/pytorch/tutorials/pull/2046)

PR #2: [Sparsity](https://github.com/pytorch/tutorials/pull/2043) -- this PR

PR #3: [Adagrad sparse semantics](https://github.com/pytorch/tutorials/pull/2047), i.e. what does MaskedTensor make easier

PR #4: Advanced semantics, e.g. NumPy differences and reduction semantics from [Old overview](https://github.com/pytorch/tutorials/pull/2042)

### CODE REVIEW

File: prototype_source/maskedtensor_sparsity.rst [hunk 1]
```diff
@@ -0,0 +1,217 @@
+Sparsity
```
Comment (mikaylagawarecki) [inline prototype_source/maskedtensor_sparsity.rst:1]: nit: title should mention masked tensor I think

File: prototype_source/maskedtensor_sparsity.rst [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        [      --,       --, 5]
+      ]
+    )
+
+Instead of using :meth:`torch.Tensor.to_sparse`, we can also create the sparse COO tensors directly,
+which brings us to a warning:
+
+.. warning::
+
+  When using a function like :meth:`MaskedTensor.to_sparse_coo`, if the user does not specify the indices
```
Comment (mikaylagawarecki) [inline prototype_source/maskedtensor_sparsity.rst:142]: The flow here feels confusing as `to_sparse_coo` wasn't used in the above code block, was this warning supposed to say when creating data/mask with `torch.Tensor.to_sparse`? Or is this trying to create a link between the behavior of tensor.to_sparse and the behavior of `MaskedTensor.to_sparse_coo`?

File: prototype_source/maskedtensor_sparsity.rst [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+-----------------------
+
+We can also construct a sparse CSR MaskedTensor using sparse CSR tensors,
+and like the example above, this results in a similar treatment under the hood.
+
+    >>> crow_indices = torch.tensor([0, 2, 4])
+    >>> col_indices = torch.tensor([0, 1, 0, 1])
+    >>> values = torch.tensor([1, 2, 3, 4])
+    >>> mask_values = torch.tensor([True, False, False, True])
+    >>>
```
Comment (mikaylagawarecki) [inline prototype_source/maskedtensor_sparsity.rst:201]: nit: extra `>>>` here and on 204

File: prototype_source/maskedtensor_sparsity.rst [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+please refer to the `sparsity tutorial <https://pytorch.org/docs/stable/sparse.html>`__ for additional details.
+
+Sparse storage formats have been proven to be powerful in a variety of ways. As a primer, the first use case
+most practitioners think about is when the majority of elements are equal to zero (a high degree of sparsity),
+but even in cases of lower sparsity, certain formats (e.g. BSR) can take advantage of substructures within a matrix.
+
+.. note::
+
+    At the moment, MaskedTensor supports COO and CSR tensors with plans to support additional formats
+    (e.g. BSR and CSC) in the future. If you have any requests for additional formats, please file a feature request!
```
Comment (svekars) [inline prototype_source/maskedtensor_sparsity.rst:14]:
```suggestion
    (such as BSR and CSC) in the future. If you have any requests for additional formats, please file a feature request `here<add link>`__!
```

File: prototype_source/maskedtensor_sparsity.rst [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+Similarly, :class:`MaskedTensor` also supports the
+`CSR (Compressed Sparse Row) <https://pytorch.org/docs/stable/sparse.html#sparse-csr-tensor>`__
+sparse tensor format. Instead of storing the tuples of the indices like sparse COO tensors, sparse CSR tensors
+aim to decrease the memory requirements by storing compressed row indices.
+In particular, a CSR sparse tensor consists of three 1-D tensors:
+
+* ``crow_indices``: array of compressed row indices with size ``(size[0] + 1,)``. This array indicates which row
+  a given entry in values lives in. The last element is the number of specified elements,
+  while crow_indices[i+1] - crow_indices[i] indicates the number of specified elements in row i.
```
Comment (svekars) [inline prototype_source/maskedtensor_sparsity.rst:82]:
```suggestion
  while `crow_indices[i+1] - crow_indices[i]` indicates the number of specified elements in row i.
```

File: prototype_source/maskedtensor_sparsity.rst [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+Appendix
+++++++++
+
+.. _sparse-coo-appendix:
+
+Sparse COO construction
+-----------------------
+
+Recall in our original example, we created a :class:`MaskedTensor` and then converted it to a sparse COO MaskedTensor
```
Comment (svekars) [inline prototype_source/maskedtensor_sparsity.rst:111]: Make `original example` a link

File: prototype_source/maskedtensor_sparsity.rst [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        [      --,       --, 5]
+      ]
+    )
+Note that ``mt`` and ``mt2`` look identical on the surface, and in the vast majority of operations, will yield the same
+result. But this brings us to a detail on the implementation:
+
+``data`` and ``mask`` -- only for sparse MaskedTensors -- can have a different number of elements (:func:`nnz`)
+**at creation**, but the indices of ``mask`` must then be a subset of the indices of ``data``. In this case,
+``data`` will assume the shape of ``mask`` by ``data = data.sparse_mask(mask)``; in other words, any of the elements
+in ``data`` that are not ``True`` in ``mask`` (i.e. not specified) will be thrown away.
```
Comment (svekars) [inline prototype_source/maskedtensor_sparsity.rst:173]:
```suggestion
in ``data`` that are not ``True`` in ``mask`` (for example, not specified) will be thrown away.
```

File: prototype_source/maskedtensor_sparsity.rst [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+      [
+        [  1.0000,       --],
+        [      --,   4.0000]
+      ]
+    )
+    >>> mt.get_data()
+    tensor(crow_indices=tensor([0, 2, 4]),
+          col_indices=tensor([0, 1, 0, 1]),
+          values=tensor([1., 2., 3., 4.]), size=(2, 2), nnz=4,
+          dtype=torch.float64, layout=torch.sparse_csr)
```
Comment (svekars) [inline prototype_source/maskedtensor_sparsity.rst:217]: Add a Conclusion section

```
Conclusion
-----------

In this tutorial we have learned. 

Add Further reading

Further reading
----------------

* Link to the next tutorial

Comment (george-qi) [reply prototype_source/maskedtensor_sparsity.rst]: done!

File: prototype_source/maskedtensor_sparsity.rst [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+        [      --,       --, 3],
+        [      --,       --, 5]
+      ]
+    )
+    >>> mt_sparse_csr.get_data()
+    tensor(crow_indices=tensor([0, 1, 2]),
+          col_indices=tensor([2, 2]),
+          values=tensor([3, 5]), size=(2, 3), nnz=2, layout=torch.sparse_csr)
+
+Appendix
```
[the remark this answers is not in the document]
Comment (george-qi) [reply prototype_source/maskedtensor_sparsity.rst]: That makes sense - added a section on operations so I think it's quite a bit longer now!

File: prototype_source/maskedtensor_sparsity.rst (hunk 2 above)
Comment (george-qi) [reply prototype_source/maskedtensor_sparsity.rst]: Ah - `to_sparse` creates a sparse COO tensor by default. I'll clarify

File: prototype_source/maskedtensor_sparsity.py [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+
+######################################################################
+# Reductions
+# ^^^^^^^^^^
+# Finally, `reductions <https://pytorch.org/docs/master/masked.html#reductions>`__ are supported:
+#
+
+print("mt:\n", mt)
+print("mt.sum():\n", mt.sum())
+print("mt.sum():\n", mt.sum(dim=1))
```
Comment (mikaylagawarecki) [inline prototype_source/maskedtensor_sparsity.py:145]: this should be `print("mt.sum(dim=1):\n", mt.sum(dim=1))`

**Review by mikaylagawarecki — APPROVED** [review]

File: prototype_source/maskedtensor_sparsity.py [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+#
+
+mt.sin()
+
+######################################################################
+# Binary
+# ^^^^^^
+# `Binary operators <https://pytorch.org/docs/master/masked.html#unary-operators>`__ are also supported, but the
+# input masks from the two masked tensors must match. For more information on why this decision was made, please
+# find our `MaskedTensor: Advanced Semantics tutorial <https://pytorch.org/tutorials/prototype/maskedtensor_advanced_semantics.html>`
```
Comment (mikaylagawarecki) [inline prototype_source/maskedtensor_sparsity.py:114]: is this missing a __ at the end?

Comment (george-qi) [reply prototype_source/maskedtensor_sparsity.py:114]: yep, thanks :)
